apiserver/facades/client/modelgeneration/modelgeneration.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelgeneration
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/facade"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/permission"
+)
+
+var logger = loggo.GetLogger("juju.apiserver.modelgeneration")
+
+// ModelGenerationAPI implements the ModelGeneration interface and is the concrete implementation
+// of the API endpoint.
+type ModelGenerationAPI struct {
+	check             *common.BlockChecker
+	authorizer        facade.Authorizer
+	apiUser           names.UserTag
+	isControllerAdmin bool
+	model             GenerationModel
+}
+
+// NewModelGenerationFacade provides the signature required for facade registration.
+func NewModelGenerationFacade(ctx facade.Context) (*ModelGenerationAPI, error) {
+	authorizer := ctx.Auth()
+	st := &modelGenerationStateShim{State: ctx.State()}
+	model, err := st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return NewModelGenerationAPI(st, authorizer, model)
+}
+
+// NewModelGenerationAPI creates a new API endpoint for dealing with model generations.
+func NewModelGenerationAPI(
+	st ModelGenerationState,
+	//resources facade.Resources,
+	authorizer facade.Authorizer,
+	m GenerationModel,
+) (*ModelGenerationAPI, error) {
+	if !authorizer.AuthClient() {
+		return nil, common.ErrPerm
+	}
+	// Since we know this is a user tag (because AuthClient is true),
+	// we just do the type assertion to the UserTag.
+	apiUser, _ := authorizer.GetAuthTag().(names.UserTag)
+	// Pretty much all of the user manager methods have special casing for admin
+	// users, so look once when we start and remember if the user is an admin.
+	isAdmin, err := authorizer.HasPermission(permission.SuperuserAccess, st.ControllerTag())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &ModelGenerationAPI{
+		//state:     st,
+		//resources: resources,
+		authorizer: authorizer,
+		//check:          common.NewBlockChecker(st),
+		isControllerAdmin: isAdmin,
+		apiUser:           apiUser,
+		model:             m,
+	}, nil
+}
+
+// authCheck checks if the user is acting on their own behalf, or if they
```

Comment:
Is this dead, or waiting for some feedback regarding which users should be authorised?

Reply:
It's dead, removed.  We should have a  discussion about what the user permissions for this should be.  Right now admin access is required.  I've added a card to Trello for the topic.